sktime/decomposition/vamp.py

```diff
@@ -0,0 +1,760 @@
+            Q[:, 0] = observables.T.dot(m_t)
+        Q[:, 1:] = observables.T.dot(self.cov_tt).dot(V)
+
+        if statistics is not None:
+            # compute covariance
+            R = np.zeros((statistics.shape[1], dim + 1))
+            if not statistics_mean_free:
+                R[:, 0] = statistics.T.dot(m_0)
+            R[:, 1:] = statistics.T.dot(self.cov_00).dot(U)
+
+        if statistics is not None:
+            # compute lagged covariance
+            return Q.dot(P).dot(R.T)
+            # TODO: discuss whether we want to return this or the transpose
+            # TODO: from MSMs one might expect to first index to refer to the statistics, here it is the other way round
+        else:
+            # compute future expectation
+            return Q.dot(P)[:, 0]
+
+    def _diagonalize(self):
+        """Performs SVD on covariance matrices and save left, right singular vectors and values in the model.
+
+        Parameters
+        ----------
+        scaling : None or string, default=None
+            Scaling to be applied to the VAMP modes upon transformation
+            * None: no scaling will be applied, variance of the singular
+              functions is 1
+            * 'kinetic map' or 'km': singular functions are scaled by
+              singular value. Note that only the left singular functions
+              induce a kinetic map.
+        """
+        L0 = spd_inv_split(self.cov_00, epsilon=self.epsilon)
+        self._rank0 = L0.shape[1] if L0.ndim == 2 else 1
+        Lt = spd_inv_split(self.cov_tt, epsilon=self.epsilon)
+        self._rankt = Lt.shape[1] if Lt.ndim == 2 else 1
+
+        W = np.dot(L0.T, self.cov_0t).dot(Lt)
+        from scipy.linalg import svd
+        A, s, BT = svd(W, compute_uv=True, lapack_driver='gesvd')
+
+        self._singular_values = s
+
+        # don't pass any values in the argument list that call _diagonalize again!!!
+        m = VAMPModel._dimension(self._rank0, self._rankt, self.dim, self._singular_values)
+
+        U = np.dot(L0, A[:, :m])
+        V = np.dot(Lt, BT[:m, :].T)
+
+        # scale vectors
+        if self.scaling is not None:
+            U *= s[np.newaxis, 0:m]  # scaled left singular functions induce a kinetic map
+            V *= s[np.newaxis, 0:m]  # scaled right singular functions induce a kinetic map wrt. backward propagator
+
+        self._U = U
+        self._V = V
+        self._svd_performed = True
+
+    def transform(self, X):
+        r"""Projects the data onto the dominant singular functions.
+
+        Parameters
+        ----------
+        X : ndarray(n, m)
+            the input data
+
+        Returns
+        -------
+        Y : ndarray(n,)
+            the projected data
+            If `right` is True, projection will be on the right singular
+            functions. Otherwise, projection will be on the left singular
+            functions.
+        """
+        # TODO: in principle get_output should not return data for *all* frames!
```

Comment:
is this still relevant?

Reply:
it is, since the last tau frames are being projected to something undefined, but I guess it works in practice.